text/3502-cargo-script.md

```diff
@@ -0,0 +1,1229 @@
+  - [`#` repl directives](https://github.com/dotnet-script/dotnet-script#repl-commands) can appear on lines following `#!`
+
+Haskell
+- [`runghc` / `runhaskell`](https://downloads.haskell.org/ghc/latest/docs/users_guide/runghc.html)
+  - Users can use the file stem (ie leave off the extension) when passing it in
+- [cabal's single-file haskel script](https://cabal.readthedocs.io/en/stable/getting-started.html#run-a-single-file-haskell-script)
+  - Command is just `cabal`, which could run into weird situations if a file has the same name as a subcommand
+  - Manifest is put in a multi-line comment that starts with `cabal:`
+  - Scripts are run with `--quiet`, regardless of which invocation is used
+  - Documented in their "Getting Started" and then documented further under `cabal run`.
+- [`stack script`](https://www.wespiser.com/posts/2020-02-02-Command-Line-Haskell.html)
+  - `stack` acts as a shortcut for use in `#!`
+  - Delegates resolver information but can be extended on the command-line
+  - Command-line flags may be specified in a multi-line comment starting with `stack script`
+
+Bash
+- `bash` to get an interactive way of entering code
+- `bash file` will run the code in `file,` searching in `PATH` if it isn't available locally
+- `./file` with `#!/usr/bin/env bash` to make standalone executables
+- `bash -c <expr>` to try out an idea right now
+- Common configuration with rc files, `--rcfile <path>`
+
+Python
+- `python` to get an interactive way of entering code
+- `python -i ...` to make other ways or running interactive
+- `python <file>` will run the file
+- `./file` with `#!/usr/bin/env python` to make standalone executables
+- `python -c <expr>` to try out an idea right now
+- Can run any file in a project (they can have their own "main") to do whitebox exploratory programming and not just blackblox
+
+Go
+- [`gorun`](https://github.com/erning/gorun/) attempts to bring that experience to a compiled language, go in this case
+  - `gorun <file>` to build and run a file
+  - Implicit garbage collection for build cache
+  - Project metadata is specified in HEREDOCs in regular code comments
+
+Perl
+- [Re-interprets the `#!`](https://stackoverflow.com/questions/38059830/how-does-perl-avoid-shebang-loops)
+
+Ruby
+- [`bundler/inline`](https://bundler.io/guides/bundler_in_a_single_file_ruby_script.html)
+  - Uses a code-block to define dependencies, making them available for use
+
+Cross-language
+- [`scriptisto`](https://github.com/igor-petruk/scriptisto)
+  - Supports any compiled language
+  - Comment-directives give build commands
+- [nix-script](https://github.com/BrianHicks/nix-script)
+  - Nix version of scriptisto, letting you use any Nix dependency
+
+See also [Single-file scripts that download their dependencies](https://dbohdan.com/scripts-with-dependencies)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Wherther `version` should may be unset or defaulted to `0.0.0`
```

Comment:
```suggestion
- Whether `version` should may be unset or defaulted to `0.0.0`
```

Reply:
if we don't set it, how would things like https://docs.rs/clap/4.4.6/clap/macro.crate_version.html react to that?

Reply:
Thats something we'll need to work out as part of the version rust-lang/cargo#9829 if we decide to do it.